## Data Analysis using Python and Pandas

#### Using the states.csv file from https://covid19india.org that cotains National as well as State-Level Covid Data for India

In [1]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

Read the CSV file using Pandas

In [2]:
covid_df = pd.read_csv('./Data/states.csv')

Data from the file is parsed and stored in a 'Data Frame' object, which is one of the core structures in Pandas for storing and working with Tabular data

In [3]:
type(covid_df)

pandas.core.frame.DataFrame

In [4]:
# Display the Data
covid_df

Date          State  Confirmed  Recovered  Deceased  Other  \
0      2020-01-30         Kerala          1          0         0      0   
1      2020-01-30          India          1          0         0      0   
2      2020-02-02         Kerala          2          0         0      0   
3      2020-02-02          India          2          0         0      0   
4      2020-02-03         Kerala          3          0         0      0   
...           ...            ...        ...        ...       ...    ...   
10630  2021-01-05        Tripura      33281      32803       382     23   
10631  2021-01-05          India   10375478    9996501    150151   4269   
10632  2021-01-05  Uttar Pradesh     589611     568892      8433      0   
10633  2021-01-05    Uttarakhand      92366      85883      1544   1222   
10634  2021-01-05    West Bengal     556384     537250      9841      0   

            Tested  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  
...            ...  
10630     585181.0  
10631  176532297.0  
10632   24594871.0  
10633    1842877.0  
10634    7266038.0  

[10635 rows x 7 columns]

Here's what we can say about the data as of now:
    - The file provides 4 main metrics for Covid19 in India and respective States
    - The metrics are: Confirmed, Recovered, Deceased and Tested
    - Data provied contains 10635 rows, starting from 30th January 2020 up to 5th January 2021
    
NOTE: These are official numbers. Actual numbers may vary based on whether cases were accurately diagnosed / reported

View basic information about the Data Frame

In [5]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10635 entries, 0 to 10634
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       10635 non-null  object 
 1   State      10635 non-null  object 
 2   Confirmed  10635 non-null  int64  
 3   Recovered  10635 non-null  int64  
 4   Deceased   10635 non-null  int64  
 5   Other      10635 non-null  int64  
 6   Tested     9907 non-null   float64
dtypes: float64(1), int64(4), object(2)
memory usage: 581.7+ KB


Notice that each column contains values of a specific type. Statistical information can be viewed for numerical columns (mean, standard deviation, min / max values) using the 'Describe' method

In [6]:
covid_df.describe()

Confirmed     Recovered       Deceased         Other        Tested
count  1.063500e+04  1.063500e+04   10635.000000  10635.000000  9.907000e+03
mean   2.136106e+05  1.881886e+05    3371.739915     45.979690  3.340063e+06
std    9.181888e+05  8.405555e+05   14286.383858    241.187968  1.367680e+07
min    0.000000e+00  0.000000e+00       0.000000     -1.000000  5.800000e+01
25%    8.670000e+02  2.820000e+02       3.000000      0.000000  5.053850e+04
50%    1.033700e+04  6.912000e+03     119.000000      0.000000  2.935910e+05
75%    1.180220e+05  9.875000e+04    1311.500000      0.000000  2.507227e+06
max    1.037548e+07  9.996501e+06  150151.000000   4269.000000  1.765323e+08

NOTE: Some of those numbers look strange. Why is that so?
If there is even one column in our data that has 'float' type values, then the describe method shows all numerical values as floats, even though we have 4 columns in our data that are 'int'. In Python, an 'int' variable can store more data than a 'float' and so converting a large int value to float results in a loss of information. Therefore, when we use the 'describe' method on this data, the columns that have large int values (Confirmed, Recovered, Tested) are converted to floats and represented this way. We will try and see if we can resolve this a bit later in this exercise.


The 'columns' property gives us a list of all columns in the data frame

In [7]:
covid_df.columns

Index(['Date', 'State', 'Confirmed', 'Recovered', 'Deceased', 'Other',
       'Tested'],
      dtype='object')

The 'shape' property gives us the number of rows and columns in the data frame

In [8]:
covid_df.shape

(10635, 7)

Here's a summary of what we have learnt in the above section:

    - pd.read_csv(): read data from a csv into a data frame object
    - .info(): view basic information about rows, columns and data types
    - .describe(): view stastical information for numerical columns
    - .columns: shows a list of all columns in the data frame
    - .shape: shows the number of rows and columns in the data frame

### Retrieving Data from a Data Frame

We are now interested in retrieving data from our data frame. To do this, it might help to understand how data is internally represented in a data frame. A data frame is a dictionary of lists (and not a list of dictionaries). The keys are the column names and values are lists/arrays containing data for the respective columns.

In [9]:
# Pandas data frames are similar to this:
covid_data_dict = {
    
    'date': ['2020-08-30', '2020-08-31', '2020-09-01'],
    'confirmed': [2000, 3000, 4000],
    'recovered': [2400, 3200, 2000],
    'deceased': [100, 93, 101],
    'tested': [10000, 10345, 12000]
}

In [10]:
# Pandas data frames are NOT as follows:
covid_data_list = [
    
    {'date': '2020-08-30', 'confirmed': 2000, 'recovered': 2400, 'deceased': 100, 'tested': 10000},
    {'date': '2020-08-31', 'confirmed': 3000, 'recovered': 3200, 'deceased': 93, 'tested': 10345},
    {'date': '2020-09-01', 'confirmed': 4000, 'recovered': 2000, 'deceased': 101, 'tested': 12000}
]

With this understanding, we can guess how to retrieve data from a data frame. We can get a list of specific values from a column using the index [ ] notation

In [11]:
covid_data_dict['recovered']

[2400, 3200, 2000]

Let's retrieve data from our main data frame

In [12]:
covid_df['Recovered']

0              0
1              0
2              0
3              0
4              0
          ...   
10630      32803
10631    9996501
10632     568892
10633      85883
10634     537250
Name: Recovered, Length: 10635, dtype: int64

Each column in the data frame is represented as a 'series', which basically is a numpy array that has additional methods and properties

In [13]:
type(covid_df['Recovered'])

pandas.core.series.Series

Let's look at how we can rename the columns of our data frame

In [14]:
covid_df.columns = ['date', 'state', 'total_confirmed', 'total_recovered', 'total_deceased', 'other', 'tested']

To view the first 5 records of the data frame, we can use the 'head' method. This also helps us verify that the column renaming was successful

In [15]:
covid_df.head()

date   state  total_confirmed  total_recovered  total_deceased  \
0  2020-01-30  Kerala                1                0               0   
1  2020-01-30   India                1                0               0   
2  2020-02-02  Kerala                2                0               0   
3  2020-02-02   India                2                0               0   
4  2020-02-03  Kerala                3                0               0   

   other  tested  
0      0     NaN  
1      0     NaN  
2      0     NaN  
3      0     NaN  
4      0     NaN

We had learnt above that a column in the data frame is basically a numpy array with additional methods and properties. Like arrays, we can retrieve a specific value in the series using the index [ ] notation

In [16]:
covid_df['total_recovered'][6000]

13186

Pandas also provies the 'at' method to retrieve a value from a row and column directly

In [17]:
covid_df.at[6000, 'total_recovered']

13186

Apart from the index notation, Pandas also allows accessing columns using '.' notation. However, this property works only when the column name does not have spaces or special characters

In [18]:
covid_df.total_recovered

0              0
1              0
2              0
3              0
4              0
          ...   
10630      32803
10631    9996501
10632     568892
10633      85883
10634     537250
Name: total_recovered, Length: 10635, dtype: int64

Also, we can pass a list of columns with the index [ ] notation to access a subset of the dataframe

In [19]:
rec_cases_df = covid_df[['date', 'total_recovered']]
rec_cases_df

date  total_recovered
0      2020-01-30                0
1      2020-01-30                0
2      2020-02-02                0
3      2020-02-02                0
4      2020-02-03                0
...           ...              ...
10630  2021-01-05            32803
10631  2021-01-05          9996501
10632  2021-01-05           568892
10633  2021-01-05            85883
10634  2021-01-05           537250

[10635 rows x 2 columns]

The new data frame 'rec_cases_df' is simply a 'snapshot' of the original data frame covid_df. Both data frames point to the same location in computer memory which effectively means that changing values in one date frame will affect the other data frame as well. This sharing of data between data frames makes data manipulation really fast in Pandas.

Sometimes though we may need to make a copy of the data frame, which can be done using the 'copy()' method

In [20]:
covid_df_copy = covid_df.copy()

In this case, the covid_df_copy data frame is separate from covid_df, and so changing values in the copy will not affect the original data frame

To access a specific row in the data frame, Pandas provides the 'loc' method

In [21]:
covid_df.loc[6000]

date               2020-08-30
state                     Goa
total_confirmed         17004
total_recovered         13186
total_deceased            183
other                       0
tested                 197216
Name: 6000, dtype: object

We had earlier seen the 'head()' method that gives us the first 5 rows from the data frame. We can pass in a number as an argument to the method to view that many rows

In [22]:
covid_df.head(20)

date      state  total_confirmed  total_recovered  total_deceased  \
0   2020-01-30     Kerala                1                0               0   
1   2020-01-30      India                1                0               0   
2   2020-02-02     Kerala                2                0               0   
3   2020-02-02      India                2                0               0   
4   2020-02-03     Kerala                3                0               0   
5   2020-02-03      India                3                0               0   
6   2020-02-14     Kerala                3                3               0   
7   2020-02-14      India                3                3               0   
8   2020-02-20     Kerala                3                3               0   
9   2020-02-20     Punjab                0                0               0   
10  2020-02-20      India                3                3               0   
11  2020-03-02      Delhi                1                0               0   
12  2020-03-02     Kerala                3                3               0   
13  2020-03-02     Punjab                0                0               0   
14  2020-03-02  Telangana                1                0               0   
15  2020-03-02      India                5                3               0   
16  2020-03-03      Delhi                1                0               0   
17  2020-03-03     Kerala                3                3               0   
18  2020-03-03     Punjab                0                0               0   
19  2020-03-03  Rajasthan                1                0               0   

    other     tested  
0       0        NaN  
1       0        NaN  
2       0        NaN  
3       0        NaN  
4       0        NaN  
5       0        NaN  
6       0        NaN  
7       0        NaN  
8       0        NaN  
9       0  3675493.0  
10      0        NaN  
11      0        NaN  
12      0        NaN  
13      0  3675493.0  
14      0        NaN  
15      0        NaN  
16      0        NaN  
17      0        NaN  
18      0  3675493.0  
19      0        NaN

The 'tail' method shows us the last 5 rows. We can pass a number argument to this method as well to view that many last rows

In [23]:
covid_df.tail()

date          state  total_confirmed  total_recovered  \
10630  2021-01-05        Tripura            33281            32803   
10631  2021-01-05          India         10375478          9996501   
10632  2021-01-05  Uttar Pradesh           589611           568892   
10633  2021-01-05    Uttarakhand            92366            85883   
10634  2021-01-05    West Bengal           556384           537250   

       total_deceased  other       tested  
10630             382     23     585181.0  
10631          150151   4269  176532297.0  
10632            8433      0   24594871.0  
10633            1544   1222    1842877.0  
10634            9841      0    7266038.0

In [24]:
covid_df.tail(10)

date          state  total_confirmed  total_recovered  \
10625  2021-01-05     Puducherry            38300            37292   
10626  2021-01-05      Rajasthan           310675           300126   
10627  2021-01-05         Sikkim             5938             5221   
10628  2021-01-05      Telangana           287993           281400   
10629  2021-01-05     Tamil Nadu           822370           802385   
10630  2021-01-05        Tripura            33281            32803   
10631  2021-01-05          India         10375478          9996501   
10632  2021-01-05  Uttar Pradesh           589611           568892   
10633  2021-01-05    Uttarakhand            92366            85883   
10634  2021-01-05    West Bengal           556384           537250   

       total_deceased  other       tested  
10625             635      0     502458.0  
10626            2719      0    5358544.0  
10627             129     95      69556.0  
10628            1554      0    7061049.0  
10629           12177      0   14502929.0  
10630             382     23     585181.0  
10631          150151   4269  176532297.0  
10632            8433      0   24594871.0  
10633            1544   1222    1842877.0  
10634            9841      0    7266038.0

We can use the 'sample' method to retrieve a random sample of rows

In [25]:
covid_df.sample(10)

date                                     state  total_confirmed  \
7326  2020-10-06                                     Assam           188902   
2810  2020-06-03                             Uttar Pradesh             8870   
2511  2020-05-26                                   Tripura              232   
4731  2020-07-26               Andaman and Nicobar Islands              318   
5963  2020-08-29  Dadra and Nagar Haveli and Daman and Diu             2308   
1818  2020-05-07                                 Telangana             1122   
2596  2020-05-29                                Chandigarh              289   
4532  2020-07-20                                    Ladakh             1195   
1226  2020-04-21               Andaman and Nicobar Islands               17   
2825  2020-06-04                                   Haryana             3281   

      total_recovered  total_deceased  other     tested  
7326           155074             778      3  3705740.0  
2810             5257             230      0   317780.0  
2511              165               0      0    22049.0  
4731              180               0      1    22049.0  
5963             1989               2     29    53527.0  
1818              693              29      0    19278.0  
2596              189               4      0     4543.0  
4532             1007               2      0    16467.0  
1226               11               0      0     1403.0  
2825             1123              24      0   132575.0

Here's a summary of what we learnt in the above section:
    - covid_df['daily_recovered']: retrieving columns as a series using the column name
    - daily_recovered[6000]: retrieving values from a series using an index
    - covid_df.at[6000, 'daily_recovered]: retrieving a single value from the data frame
    - covid_df.copy(): create a deep copy of the data frame
    - covid_df.loc[6000]: retrieving a row or range of rows from the data frame
    - head, tail and sample: methods used to retrieve several rows from the data frame

### Analyzing Data from a Data Frame

We now begin asking questions about our data.

Our data frame contains India + States data, resulting in 10635 rows. To keep things simple, we can choose to work with a subset of the data and perform analysis for a specific State. For this exercise, we will extract and work with India-level data

Create a new dataframe for India by querying the data and extracting all rows where 'state' is 'India'

In [26]:
covid_ind_df = covid_df[covid_df['state'] == 'India']

View the extracted data

In [27]:
covid_ind_df

date  state  total_confirmed  total_recovered  total_deceased  \
1      2020-01-30  India                1                0               0   
3      2020-02-02  India                2                0               0   
5      2020-02-03  India                3                0               0   
7      2020-02-14  India                3                3               0   
10     2020-02-20  India                3                3               0   
...           ...    ...              ...              ...             ...   
10487  2021-01-01  India         10306469          9905570          149255   
10523  2021-01-02  India         10324613          9926473          149471   
10559  2021-01-03  India         10341291          9946131          149686   
10595  2021-01-04  India         10357569          9975340          149886   
10631  2021-01-05  India         10375478          9996501          150151   

       other       tested  
1          0          NaN  
3          0          NaN  
5          0          NaN  
7          0          NaN  
10         0          NaN  
...      ...          ...  
10487   4217  173941958.0  
10523   4233  174900083.0  
10559   4246  175636061.0  
10595   4255  176532297.0  
10631   4269  176532297.0  

[315 rows x 7 columns]

In [28]:
covid_ind_df.shape

(315, 7)

So we now have 315 rows of data to work with instead of 10635.
Before proceeding with the analysis, we need to decide if we want to continue working with the data that has been provided or if we need new information / features. The data already gives us the 'total' count of confirmed, recovered and deceased cases. It would be a good idea to have a 'daily' count for these columns. We will start by creating three new columns that will store this information

In [29]:
covid_ind_df['daily_confirmed'] = 0
covid_ind_df['daily_recovered'] = 0
covid_ind_df['daily_deceased'] = 0

covid_ind_df.head()

date  state  total_confirmed  total_recovered  total_deceased  \
1   2020-01-30  India                1                0               0   
3   2020-02-02  India                2                0               0   
5   2020-02-03  India                3                0               0   
7   2020-02-14  India                3                3               0   
10  2020-02-20  India                3                3               0   

    other  tested  daily_confirmed  daily_recovered  daily_deceased  
1       0     NaN                0                0               0  
3       0     NaN                0                0               0  
5       0     NaN                0                0               0  
7       0     NaN                0                0               0  
10      0     NaN                0                0               0

In [30]:
covid_ind_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 315 entries, 1 to 10631
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   date             315 non-null    object 
 1   state            315 non-null    object 
 2   total_confirmed  315 non-null    int64  
 3   total_recovered  315 non-null    int64  
 4   total_deceased   315 non-null    int64  
 5   other            315 non-null    int64  
 6   tested           299 non-null    float64
 7   daily_confirmed  315 non-null    int64  
 8   daily_recovered  315 non-null    int64  
 9   daily_deceased   315 non-null    int64  
dtypes: float64(1), int64(7), object(2)
memory usage: 27.1+ KB


The three new columns are of the 'int' type and have been initialized with 0 for all rows. We will need to compute values for these columns. Let's start with the daily_confirmed

In [32]:
covid_ind_df['daily_confirmed'] = covid_ind_df['total_confirmed'] - covid_ind_df['total_confirmed'].shift(1)
covid_ind_df['daily_confirmed']

1            NaN
3            1.0
5            1.0
7            0.0
10           0.0
          ...   
10487    20159.0
10523    18144.0
10559    16678.0
10595    16278.0
10631    17909.0
Name: daily_confirmed, Length: 315, dtype: float64

What did we just do? We took value of the 'total_confirmed' from each row and subtracted it from the value of the previous row (using the 'shift' method of python). This gives us the daily increase / decrease value for confirmed cases, which is then stored in the 'daily_confirmed' column for each row. Let's compute values for 'daily_recovered' and 'daily_deceased'

In [33]:
covid_ind_df['daily_recovered'] = covid_ind_df['total_recovered'] - covid_ind_df['total_recovered'].shift(1)
covid_ind_df['daily_deceased'] = covid_ind_df['total_deceased'] - covid_ind_df['total_deceased'].shift(1)

In [34]:
covid_ind_df['daily_recovered']

1            NaN
3            0.0
5            0.0
7            3.0
10           0.0
          ...   
10487    23838.0
10523    20903.0
10559    19658.0
10595    29209.0
10631    21161.0
Name: daily_recovered, Length: 315, dtype: float64

In [35]:
covid_ind_df['daily_deceased']

1          NaN
3          0.0
5          0.0
7          0.0
10         0.0
         ...  
10487    237.0
10523    216.0
10559    215.0
10595    200.0
10631    265.0
Name: daily_deceased, Length: 315, dtype: float64

This is how our data frame looks now

In [36]:
covid_ind_df.head(10)

date  state  total_confirmed  total_recovered  total_deceased  \
1   2020-01-30  India                1                0               0   
3   2020-02-02  India                2                0               0   
5   2020-02-03  India                3                0               0   
7   2020-02-14  India                3                3               0   
10  2020-02-20  India                3                3               0   
15  2020-03-02  India                5                3               0   
21  2020-03-03  India                6                3               0   
28  2020-03-04  India               28                3               0   
36  2020-03-05  India               30                3               0   
44  2020-03-06  India               31                3               0   

    other  tested  daily_confirmed  daily_recovered  daily_deceased  
1       0     NaN              NaN              NaN             NaN  
3       0     NaN              1.0              0.0             0.0  
5       0     NaN              1.0              0.0             0.0  
7       0     NaN              0.0              3.0             0.0  
10      0     NaN              0.0              0.0             0.0  
15      0     NaN              2.0              0.0             0.0  
21      0     NaN              1.0              0.0             0.0  
28      0     NaN             22.0              0.0             0.0  
36      0     NaN              2.0              0.0             0.0  
44      0     NaN              1.0              0.0             0.0

Great. Let's create another new column that will store the number of 'active' cases (remember we are creating all these columns for India-level data)

In [37]:
covid_ind_df['active'] = covid_ind_df['total_confirmed'] - (covid_ind_df['total_recovered'] + covid_ind_df['total_deceased']
                                                           + covid_ind_df['other'])
covid_ind_df['active']

1             1
3             2
5             3
7             0
10            0
          ...  
10487    247427
10523    244436
10559    241228
10595    228088
10631    224557
Name: active, Length: 315, dtype: int64

In [38]:
covid_ind_df.head(10)

date  state  total_confirmed  total_recovered  total_deceased  \
1   2020-01-30  India                1                0               0   
3   2020-02-02  India                2                0               0   
5   2020-02-03  India                3                0               0   
7   2020-02-14  India                3                3               0   
10  2020-02-20  India                3                3               0   
15  2020-03-02  India                5                3               0   
21  2020-03-03  India                6                3               0   
28  2020-03-04  India               28                3               0   
36  2020-03-05  India               30                3               0   
44  2020-03-06  India               31                3               0   

    other  tested  daily_confirmed  daily_recovered  daily_deceased  active  
1       0     NaN              NaN              NaN             NaN       1  
3       0     NaN              1.0              0.0             0.0       2  
5       0     NaN              1.0              0.0             0.0       3  
7       0     NaN              0.0              3.0             0.0       0  
10      0     NaN              0.0              0.0             0.0       0  
15      0     NaN              2.0              0.0             0.0       2  
21      0     NaN              1.0              0.0             0.0       3  
28      0     NaN             22.0              0.0             0.0      25  
36      0     NaN              2.0              0.0             0.0      27  
44      0     NaN              1.0              0.0             0.0      28

We do not need the 'other' column anymore so we can remove it from the data frame using the 'drop' method

In [39]:
covid_ind_df = covid_ind_df.drop(['other'], axis = 1)
covid_ind_df.head()

date  state  total_confirmed  total_recovered  total_deceased  \
1   2020-01-30  India                1                0               0   
3   2020-02-02  India                2                0               0   
5   2020-02-03  India                3                0               0   
7   2020-02-14  India                3                3               0   
10  2020-02-20  India                3                3               0   

    tested  daily_confirmed  daily_recovered  daily_deceased  active  
1      NaN              NaN              NaN             NaN       1  
3      NaN              1.0              0.0             0.0       2  
5      NaN              1.0              0.0             0.0       3  
7      NaN              0.0              3.0             0.0       0  
10     NaN              0.0              0.0             0.0       0

The axis = 1 indicates that we are dropping a column

The first step (usually) in data analysis is to check if we have any null / missing values in our data frame. We can use the 'isnull().sum()' expression for this

In [40]:
covid_ind_df.isnull().sum()

date                0
state               0
total_confirmed     0
total_recovered     0
total_deceased      0
tested             16
daily_confirmed     1
daily_recovered     1
daily_deceased      1
active              0
dtype: int64

We have 16 null values in the 'tested' column and 1 each in three new columns we had created earlier

Pandas provides the 'isna()' method to retrieve rows that have null values from a specified column

In [41]:
covid_ind_df[covid_ind_df['tested'].isna()]

date  state  total_confirmed  total_recovered  total_deceased  \
1    2020-01-30  India                1                0               0   
3    2020-02-02  India                2                0               0   
5    2020-02-03  India                3                0               0   
7    2020-02-14  India                3                3               0   
10   2020-02-20  India                3                3               0   
15   2020-03-02  India                5                3               0   
21   2020-03-03  India                6                3               0   
28   2020-03-04  India               28                3               0   
36   2020-03-05  India               30                3               0   
44   2020-03-06  India               31                3               0   
54   2020-03-07  India               34                3               0   
64   2020-03-08  India               39                3               0   
77   2020-03-09  India               48                3               0   
90   2020-03-10  India               63                3               0   
103  2020-03-11  India               71                3               0   
117  2020-03-12  India               81                3               0   

     tested  daily_confirmed  daily_recovered  daily_deceased  active  
1       NaN              NaN              NaN             NaN       1  
3       NaN              1.0              0.0             0.0       2  
5       NaN              1.0              0.0             0.0       3  
7       NaN              0.0              3.0             0.0       0  
10      NaN              0.0              0.0             0.0       0  
15      NaN              2.0              0.0             0.0       2  
21      NaN              1.0              0.0             0.0       3  
28      NaN             22.0              0.0             0.0      25  
36      NaN              2.0              0.0             0.0      27  
44      NaN              1.0              0.0             0.0      28  
54      NaN              3.0              0.0             0.0      31  
64      NaN              5.0              0.0             0.0      36  
77      NaN              9.0              0.0             0.0      45  
90      NaN             15.0              0.0             0.0      60  
103     NaN              8.0              0.0             0.0      68  
117     NaN             10.0              0.0             0.0      78

View the first 20 rows 

In [42]:
covid_ind_df.head(20)

date  state  total_confirmed  total_recovered  total_deceased  \
1    2020-01-30  India                1                0               0   
3    2020-02-02  India                2                0               0   
5    2020-02-03  India                3                0               0   
7    2020-02-14  India                3                3               0   
10   2020-02-20  India                3                3               0   
15   2020-03-02  India                5                3               0   
21   2020-03-03  India                6                3               0   
28   2020-03-04  India               28                3               0   
36   2020-03-05  India               30                3               0   
44   2020-03-06  India               31                3               0   
54   2020-03-07  India               34                3               0   
64   2020-03-08  India               39                3               0   
77   2020-03-09  India               48                3               0   
90   2020-03-10  India               63                3               0   
103  2020-03-11  India               71                3               0   
117  2020-03-12  India               81                3               0   
131  2020-03-13  India               91                3               1   
146  2020-03-14  India              102                3               1   
161  2020-03-15  India              112                3               1   
178  2020-03-16  India              126                3               1   

     tested  daily_confirmed  daily_recovered  daily_deceased  active  
1       NaN              NaN              NaN             NaN       1  
3       NaN              1.0              0.0             0.0       2  
5       NaN              1.0              0.0             0.0       3  
7       NaN              0.0              3.0             0.0       0  
10      NaN              0.0              0.0             0.0       0  
15      NaN              2.0              0.0             0.0       2  
21      NaN              1.0              0.0             0.0       3  
28      NaN             22.0              0.0             0.0      25  
36      NaN              2.0              0.0             0.0      27  
44      NaN              1.0              0.0             0.0      28  
54      NaN              3.0              0.0             0.0      31  
64      NaN              5.0              0.0             0.0      36  
77      NaN              9.0              0.0             0.0      45  
90      NaN             15.0              0.0             0.0      60  
103     NaN              8.0              0.0             0.0      68  
117     NaN             10.0              0.0             0.0      78  
131  6500.0             10.0              0.0             1.0      87  
146  6500.0             11.0              0.0             0.0      95  
161  6500.0             10.0              0.0             0.0     105  
178  6500.0             14.0              0.0             0.0     119

It looks like the first 16 rows of our data frame has 'NaN' in the 'tested' column. Missing or null values have to be effectively handled as their presence leads to incorrect analysis. If we observe our data, it looks like we can safely delete these rows from the data frame and still have enough information to perform data analysis. 

NOTE: Deleting rows is not the only way to handle null values. If your dataset is large and has significant number of null values, then deletion may result in loss of critical information. If numerical columns have null values, then a good approach is to fill the empty cells with either the 'mean' or 'median' values of the column. There may be several other ways also to hande null values, depending on the dataset and what kind of analysis needs to be done. Handling null values therefore needs to be evaluated and executed based on the context of the data being analyzed

We can delete all rows having null values using the 'dropna()' method

In [43]:
covid_ind_df = covid_ind_df.dropna()
covid_ind_df.head(20)

date  state  total_confirmed  total_recovered  total_deceased  \
131  2020-03-13  India               91                3               1   
146  2020-03-14  India              102                3               1   
161  2020-03-15  India              112                3               1   
178  2020-03-16  India              126                3               1   
196  2020-03-17  India              146                3               1   
215  2020-03-18  India              171                3               1   
237  2020-03-19  India              198                3               1   
260  2020-03-20  India              256                4               1   
283  2020-03-21  India              334                4               1   
307  2020-03-22  India              403                4               1   
331  2020-03-23  India              497                4               1   
356  2020-03-24  India              571                6               1   
383  2020-03-25  India              657                6               1   
411  2020-03-26  India              730                6               2   
439  2020-03-27  India              883                8               3   
467  2020-03-28  India             1019                8               3   
495  2020-03-29  India             1139               99              28   
523  2020-03-30  India             1326              141              41   
553  2020-03-31  India             1635              160              47   
583  2020-04-01  India             2059              169              58   

      tested  daily_confirmed  daily_recovered  daily_deceased  active  
131   6500.0             10.0              0.0             1.0      87  
146   6500.0             11.0              0.0             0.0      95  
161   6500.0             10.0              0.0             0.0     105  
178   6500.0             14.0              0.0             0.0     119  
196   6500.0             20.0              0.0             0.0     139  
215  13125.0             25.0              0.0             0.0     164  
237  14175.0             27.0              0.0             0.0     191  
260  15404.0             58.0              1.0             0.0     248  
283  16911.0             78.0              0.0             0.0     326  
307  18127.0             69.0              0.0             0.0     395  
331  20707.0             94.0              0.0             0.0     489  
356  22694.0             74.0              2.0             0.0     561  
383  25144.0             86.0              0.0             0.0     647  
411  27688.0             73.0              0.0             1.0     719  
439  27688.0            153.0              2.0             1.0     869  
467  27688.0            136.0              0.0             0.0    1005  
495  27688.0            120.0             91.0            25.0    1009  
523  38442.0            187.0             42.0            13.0    1141  
553  42788.0            309.0             19.0             6.0    1425  
583  47951.0            424.0              9.0            11.0    1829

Let's verify that all null rows are gone

In [46]:
covid_ind_df.isnull().sum()

date               0
state              0
total_confirmed    0
total_recovered    0
total_deceased     0
tested             0
daily_confirmed    0
daily_recovered    0
daily_deceased     0
active             0
dtype: int64

We now begin data analysis by asking some basic questions

(Q) What is the total number of confirmed covid cases in India?

Our data frame contains a 'total_confirmed' column, so to answer the above question we can use the 'max' method to retrieve the value from this column and store it in a variable

In [51]:
total_confirmed_cases = covid_ind_df['total_confirmed'].max()
total_confirmed_cases

10375478

(Q) What is the total number of recovered covid cases in India?

In [52]:
total_recovered_cases = covid_ind_df['total_recovered'].max()
total_recovered_cases

9996501

(Q) What is the recovery rate? (ratio of total reported recovered cases to the total reported confirmed cases)

In [54]:
recovery_rate = int((total_recovered_cases / total_confirmed_cases) * 100)
print("The recovery rate for India is {} %".format(recovery_rate))

The recovery rate for India is 96 %


(Q) What is the total number of deaths due to covid in India?

In [55]:
total_deaths = covid_ind_df['total_deceased'].max()
total_deaths

150151

(Q) What is the case fatality rate? (ratio of total reported deaths to total reported confirmed cases)

In [61]:
cfr = (total_deaths / total_confirmed_cases) * 100
print("The case fatality rate of India is {:.2f} %".format(cfr))

The case fatality rate of India is 1.45 %


(Q) What is the overall number of tests conducted?

In [58]:
total_tests = covid_ind_df['tested'].max()
total_tests

176532297.0

(Q) What is the overall test positivity rate? (ratio of total tests to total reported confirmed cases)

In [62]:
tpr = (total_confirmed_cases / total_tests) * 100
print("The overall test positivity rate for India is {:.2f} %".format(tpr))

The overall test positivity rate for India is 5.88 %
